src/treelearner/monotone_constraints.hpp

```diff
@@ -26,24 +28,59 @@ struct ConstraintEntry {
   void UpdateMin(double new_min) { min = std::max(new_min, min); }
 
   void UpdateMax(double new_max) { max = std::min(new_max, max); }
+
+  bool UpdateMinAndReturnBoolIfChanged(double new_min) {
+    if (new_min > min) {
+      min = new_min;
+      return true;
+    }
+    return false;
+  }
+
+  bool UpdateMaxAndReturnBoolIfChanged(double new_max) {
+    if (new_max < max) {
+      max = new_max;
+      return true;
+    }
+    return false;
+  }
 };
 
-template <typename ConstraintEntry>
-class LeafConstraints {
+class LeafConstraintsBase {
  public:
-  explicit LeafConstraints(int num_leaves) : num_leaves_(num_leaves) {
+  virtual ~LeafConstraintsBase(){};
+  virtual const ConstraintEntry &Get(int leaf_idx) const = 0;
+  virtual void Reset() = 0;
+  virtual void BeforeSplit(const Tree *tree, int leaf, int new_leaf,
+                           int8_t monotone_type) = 0;
+  virtual std::vector<int> Update(
+      const Tree *tree, bool is_numerical_split,
+      int leaf, int new_leaf, int8_t monotone_type, double right_output,
+      double left_output, int split_feature, const SplitInfo &split_info,
+      const std::vector<SplitInfo> &best_split_per_leaf) = 0;
+
+  inline static LeafConstraintsBase *Create(const Config *config, int num_leaves);
+};
+
+class BasicLeafConstraints : public LeafConstraintsBase {
+ public:
+  explicit BasicLeafConstraints(int num_leaves) : num_leaves_(num_leaves) {
     entries_.resize(num_leaves_);
   }
 
-  void Reset() {
-    for (auto& entry : entries_) {
+  void Reset() override {
+    for (auto &entry : entries_) {
       entry.Reset();
     }
   }
 
-  void UpdateConstraints(bool is_numerical_split, int leaf, int new_leaf,
-                         int8_t monotone_type, double right_output,
-                         double left_output) {
+  void BeforeSplit(const Tree *, int, int, int8_t) override {}
+
+  std::vector<int> Update(const Tree *,
+                          bool is_numerical_split, int leaf, int new_leaf,
+                          int8_t monotone_type, double right_output,
+                          double left_output, int, const SplitInfo &,
+                          const std::vector<SplitInfo> &) override {
```

Comment:
sorry, I use the wrong style config. for the reference and pointers, we should use `T& x` and `T* x`, rather than `T &x` and `T *x`.